docs: schema change management (#2318)
graphql/graphql.github.io: src/pages/learn/governance-versioning.mdx

}
```

This example adds a `createdAt` field to the `User` type. Existing queries that request `id` and `email` continue working exactly as before. New clients can request `createdAt` whenever they're ready to use it.

When adding new fields, make them nullable or provide sensible defaults unless you have certainty they'll always have values. Nullable fields let you return null for older data that lacks the new information, maintaining backward compatibility.

### Handle optional arguments carefully

Adding optional arguments to fields is generally safe, but requires default behavior that matches how the field worked before the argument existed.

```graphql
type Query {
  products(
    first: Int = 20
    sortBy: ProductSort = POPULARITY
  ): [Product!]!
}
```

This example adds a `sortBy` argument to an existing products query. The default value, `POPULARITY` ensures the query behaves identically for clients that don't provide the argument. New clients can specify different sorting when needed.

To implement this safely, your resolver must handle the argument being absent and provide behavior that matches existing client expectations.

## Identify breaking changes

> Breaking changes modify or remove existing schema elements in ways that cause previously valid queries to fail or return different data.

Reviewer: ```suggestion
A breaking change is any change that breaks the server/client contract. This happens when previously valid operations become invalid or when the shape of the returned data changes.

```

Common breaking changes include:

- Removing fields or types
- Renaming fields or types
- Changing field types, such as `String` to `Int`
- Removing or renaming enum values
- Making optional arguments required

> - Changing argument types

Reviewer: ```suggestion
- Changing argument types
- Making a non-null field nullable
```

**Removing a field:**

```graphql
# Before
type User {
  id: ID!
  name: String!
  email: String!
}

# After — queries requesting `email` will fail
type User {
  id: ID!
  name: String!
}
```

**Changing a field type:**

```graphql
# Before
type Product {
  id: ID!
  price: Float!
}

# After — clients expecting Float receive Money instead
type Product {
  id: ID!
  price: Money!
}
```

> **Making a nullable field non-null:**

Reviewer: ```suggestion
**Making a non-null field nullable:**
```

```graphql
# Before
type Order {
  id: ID!
  discount: Float
}

# After — queries might receive errors if discount is null
type Order {
  id: ID!
  discount: Float!
}
```

> These examples show schema changes that break existing queries. Clients requesting the removed `email` field receive errors. Clients expecting a `Float` for `price` get a `Money` object instead. Queries relying on `null` discounts fail validation.

Reviewer: ```suggestion
These examples show schema changes that break existing queries. Clients requesting the removed `email` field receive validation errors. Clients expecting a `Float` for `price` get a `Money` object instead. Queries relying on `non-null` discounts may crash.
```

    context.metrics.recordDeprecatedFieldUsage({
      field: fieldName,
      client: context.clientId,
      timestamp: Date.now()
    });
    
    // Return the actual result
    return resolver(parent, args, context, info);
  };
}
```

This example wraps resolvers for deprecated fields to track usage. It records which client accessed the deprecated field so you can identify who needs to migrate.

To track effectively, store deprecated field usage with client identifiers in your metrics system. Query this data regularly to identify clients that haven't migrated and contact those teams directly when deprecation deadlines approach.

### Remove after migration completes

Remove deprecated elements only when usage drops to acceptable levels or the deadline passes. For critical systems, wait until usage reaches zero. For less critical fields, you might remove after usage drops below a threshold appropriate for your context.

Before removing, send final warnings to any remaining clients with a specific deadline and offer migration assistance if needed. After removal, monitor for errors that might indicate missed clients.

## Handle dangerous changes

> Some changes appear safe but can cause subtle issues. These "dangerous" changes don't break the schema structurally but might break client logic.

Reviewer: ```suggestion
Some changes appear safe but can cause subtle issues. These "dangerous" changes don't break the schema structurally and [robust applications](./robust-applications) can handle most of them but caution must be exercised.
```
